### PR TITLE
Add CUDA pool, scheduler, and other helper functionality

### LINKED ISSUE #13: Add support for cuBLAS/cuSOLVER to CUDA sender/receiver functionality

### LINKED ISSUE #14: Add CUDA stream pool, scheduler, and various helpers for dealing with CUDA

### PR DESCRIPTION
**Pull request by msimberg**
Fixes #13 and fixes #14.

Adds:
- `cuda_stream`/`cublas_handle`/`cusolver_handle` wrappers that give RAII streams/handles
- `cuda_pool`: a pool of CUDA streams (normal and high priority)
- `cuda_scheduler`: a context for doing CUDA work based on `cuda_pool` (this is a P2300 scheduler)
- renamed and extended `transform_stream` to `then_with_stream`, `then_with_cublas`, and `then_with_cusolver` to launch kernels with streams and handles
- `bulk` specialization for `cuda_scheduler`

Any other sender/receiver functionality is for later. The above functionality is what's required by DLA-Future.

The main difference to the DLA-Future version is that the pool of streams is shared by all worker threads and the handles are `thread_local`, not artificially thread local through a vector of handles of size `pika::get_worker_thread_count()`. This was done keeping in mind that this functionality could be used without the runtime. This PR still relies on the `pika` scheduler doing the polling. However, this PR does not depend on the `pika` runtime being started when creating the `cuda_pool`. Making the polling independent of the runtime will be done if/when needed.

Currently a draft. Still to do:
- [x] rebase on `main` (to integrate removal of `local` namespace etc.)
- [x] revisit cuBLAS/cuSOLVER linking (currently just hacked together, does it need to be cleaned up)
- [x] check if move/copy semantics are ok for various helpers
- [x] keep only `thread_local` cuBLAS/cuSOLVER support (i.e. remove the alternative version with cuBLAS/cuSOLVER handles kept in the CUDA pool, the former seems to work at least as well as the latter and is simpler)
- [x] revisit the APIs for `cublas_handle`/`cusolver_handle` (at least the constructors)
- [x] ~keep CUDA 11.2 requirement? (only required by tests for `cuda{Malloc,Free}Async` because it's convenient, but it's not a must)~ kept the requirement at 11.0

Apologies for one big commit. I did this because this was previously targeted at HPX and I squashed it to stay sane when moving it over here.

### CODE REVIEW

**Comment by msimberg** [comment]
test cscs

**Comment by msimberg** [comment]
Thanks to @aurianer's suggestion, I've made cuBLAS/cuSOLVER required again to simplify the CMake handling for that.

The xsolver layer currently only supports cuSOLVER, not hipsolver, see #48.

**Comment by msimberg** [comment]
test cscs

File: .jenkins/cscs/env-clang-cuda.sh [hunk 1]
```diff
@@ -10,7 +10,7 @@ export CXX_STD="17"
 export HWLOC_ROOT="${APPS_ROOT}/hwloc-2.0.3-gcc-8.3.0"
 
 module load daint-gpu
-module load cudatoolkit/11.0.2_3.38-8.1__g5b73779
+module load cudatoolkit/11.2.0_3.39-2.1__gf93aa1c
```
Comment (msimberg) [inline .jenkins/cscs/env-clang-cuda.sh:13]: Note to self: change this or the gcc build back to 11.0.

File: libs/pika/async_cuda/include/pika/async_cuda/cublas_handle.hpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    /// RAII wrapper for a cuBLAS handle.
+    class cublas_handle
+    {
+    private:
+        int device;
+        cudaStream_t stream{};
+        cublasHandle_t handle{};
+
+        using mutex_type = pika::lcos::local::spinlock;
+        mutable mutex_type mtx;
```
Comment (msimberg) [inline libs/pika/async_cuda/include/pika/async_cuda/cublas_handle.hpp:29]: Note to self: remove!?

File: libs/pika/async_cuda/include/pika/async_cuda/cuda_pool.hpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            pool_data& operator=(pool_data&&) = delete;
+            pool_data& operator=(pool_data const&) = delete;
+        };
+
+        std::shared_ptr<pool_data> data;
+
+    public:
+        PIKA_EXPORT explicit cuda_pool(int device = 0,
+            std::size_t num_normal_priority_streams = 64,
+            std::size_t num_high_priority_streams = 64);
```
Comment (msimberg) [inline libs/pika/async_cuda/include/pika/async_cuda/cuda_pool.hpp:66]: This seems to be a better default for DLA-Future, but is it a good default for everyone?
```suggestion
        PIKA_EXPORT explicit cuda_pool(int device = 0,
            std::size_t num_normal_priority_streams = 64,
            std::size_t num_high_priority_streams = 1);
```

File: libs/pika/async_cuda/include/pika/async_cuda/cublas_handle.hpp (hunk 2 above)
Comment (msimberg) [reply libs/pika/async_cuda/include/pika/async_cuda/cublas_handle.hpp:29]: Yes, removed.

File: libs/pika/async_cuda/tests/unit/cuda_scheduler.cu [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+    {
+        // This test initializes the thread_pool_scheduler with nullptr only to
+        // avoid it trying to get a thread pool through the default thread pool
+        // handler which is not installed in this test (the HPX runtime is not
+        // started). The thread pool is never accessed.
+        // auto s = ex::schedule(sched) |
+        //     cu::then_with_cusolver([](cusolverDnHandle_t) {}) |
+        //     ex::transfer(ex::thread_pool_scheduler{nullptr});
+        // CHECK_NOT_CUDA_COMPLETION_SCHEDULER(s);
```
Comment (msimberg) [inline libs/pika/async_cuda/tests/unit/cuda_scheduler.cu:76]: Note to self: this needs to be uncommented...

File: libs/pika/async_cuda/tests/unit/cuda_scheduler.cu [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    }
```
Comment (msimberg) [inline libs/pika/async_cuda/tests/unit/cuda_scheduler.cu:77]: Note to self: add tests to check that priorities are taken into account when scheduling with a `cuda_scheduler`.

**Comment by msimberg** [comment]
test cscs

File: libs/pika/async_cuda/include/pika/async_cuda/then_with_stream.hpp [hunk 6]
```diff
@@ -0,0 +1,627 @@
...
+        template <typename F>
+        constexpr PIKA_FORCEINLINE auto operator()(F&& f) const -> decltype(
+            pika::execution::experimental::detail::partial_algorithm<
+                then_with_cusolver_t, F>{PIKA_FORWARD(F, f)})
+        {
+            return pika::execution::experimental::detail::partial_algorithm<
+                then_with_cusolver_t, F>{PIKA_FORWARD(F, f)};
+        }
+    } then_with_cusolver{};
+#endif
```
Comment (msimberg) [inline libs/pika/async_cuda/include/pika/async_cuda/then_with_stream.hpp:626]: Note to self: Move the generic `then_with_anything_cuda_thing` from DLA-Future into here as well.

File: libs/pika/async_cuda/tests/unit/cuda_bulk.cu [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+#endif
+    };
+
+    // Integral shape
+    for (element_type n : {1, 42, 10007})
+    {
+        std::vector<element_type> host_vector(n, 0);
+        element_type* device_ptr = nullptr;
+
+        // We capture the and the size into the lambda. The proper way would be
```
Comment (aurianer) [inline libs/pika/async_cuda/tests/unit/cuda_bulk.cu:60]: missing word?

File: libs/pika/async_cuda/tests/unit/cuda_bulk.cu [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+    // Range
+    for (element_type n : {1, 42, 10007})
+    {
+        using element_type = int;
+
+        std::vector<element_type> host_vector(n, 0);
+        element_type* device_ptr = nullptr;
+
+        // We capture the and pointer the size into the lambda. The proper way
```
Comment (aurianer) [inline libs/pika/async_cuda/tests/unit/cuda_bulk.cu:88]: same?

**Review by aurianer — APPROVED** [review]
LGTM thanks a lot!

File: libs/pika/async_cuda/include/pika/async_cuda/cuda_future.hpp [hunk 9]
```diff
@@ -95,12 +95,13 @@ namespace pika { namespace cuda { namespace experimental {
...
                         {
-                            fdp->set_exception(
-                                std::make_exception_ptr(cuda_exception(
-                                    std::string(
-                                        "cuda function returned error code :") +
-                                        cudaGetErrorString(status),
-                                    status)));
+                            auto msg =
+                                std::string(
+                                    "CUDA function returned error code ") +
```
[the remark this answers is not in the document]
Comment (msimberg) [reply libs/pika/async_cuda/include/pika/async_cuda/cuda_future.hpp:100]: Good idea!

File: libs/pika/async_cuda/tests/unit/cuda_bulk.cu (hunk 7 above)
Comment (msimberg) [reply libs/pika/async_cuda/tests/unit/cuda_bulk.cu:60]: Definitely... same for the other one.

**Comment by msimberg** [comment]
test cscs

File: libs/pika/async_cuda/include/pika/async_cuda/cuda_pool.hpp (hunk 3 above)
Comment (msimberg) [reply libs/pika/async_cuda/include/pika/async_cuda/cuda_pool.hpp:66]: I'm keeping this as it is. It seems very algorithm-specific what number of streams is the best. At least 64/64 gives enough concurrency for submitted work. If we come to some more universal "best parameters" we can still update this. Also, they're only defaults after all.

**Comment by msimberg** [comment]
test cscs

**Comment by msimberg** [comment]
bors merge

**Pull request merged**